go/mysql/conn.go

```diff
@@ -163,24 +164,57 @@ type Conn struct {
 // bufPool is used to allocate and free buffers in an efficient way.
 var bufPool = bucketpool.New(connBufferSize, MaxPacketSize)
 
+// writersPool is used for pooling bufio.Writer objects.
+var writersPool = sync.Pool{New: func() interface{} { return bufio.NewWriterSize(nil, connBufferSize) }}
+
 // newConn is an internal method to create a Conn. Used by client and server
 // side for common creation code.
 func newConn(conn net.Conn) *Conn {
 	return &Conn{
-		conn: conn,
+		conn:           conn,
+		bufferedReader: bufio.NewReaderSize(conn, connBufferSize),
+		sequence:       0,
+	}
+}
 
-		reader:   bufio.NewReaderSize(conn, connBufferSize),
-		writer:   bufio.NewWriterSize(conn, connBufferSize),
-		sequence: 0,
+// startWriterBuffering starts using buffered writes. This should
+// be terminated by a call to flush.
+func (c *Conn) startWriterBuffering() {
+	c.bufferedWriter = writersPool.Get().(*bufio.Writer)
+	c.bufferedWriter.Reset(c.conn)
+}
+
+// flush flushes the written data to the socket.
+// This must be called to terminate startBuffering.
+func (c *Conn) flush() error {
+	if c.bufferedWriter == nil {
+		return nil
 	}
+
+	defer func() {
```

Comment:
I think defer here adds to the complexity.
Can we just
```
err := c.bufferedWriter.Flush()
...
return err
```

Reply:
The defer protects us from accidental panics. It's the recommended pattern unless there is a specific reason like performance, etc.